decision_service/dsclientlib/logger/ds_async_batcher.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "ds_concurrent_queue.h"
+#include "ds_api_status.h"
+
+#include <memory>
+#include <string>
+#include "../error_callback_fn.h"
+#include <thread>
+
+namespace decision_service {
+  class error_callback_fn;
+
+	// This class takes uses a queue and a background thread to accumulate events, and send them by batch asynchronously.
+	// A batch is shipped with TSender::send(data)
+	template <typename TSender>
+	class async_batcher {
+
+	public:
+
+		int append(const std::string& evt, api_status* status = nullptr)
+		{
+			if (_queue.size() < _queue_max_size) {
+				_queue.push(evt);
+				return error_code::success;
+			}
+
+			//report errors
+			api_status::try_update(status, error_code::background_queue_overflow, "dropped event: " + evt);
+			return error_code::background_queue_overflow;
+		}
+
+	private:
+		void timer()//the timer triggers a queue flush (run in background)
+		{
+			while (_thread_is_running)
+			{
+				std::this_thread::sleep_for(std::chrono::milliseconds(_batch_timeout_ms));
+				flush();
+			}
+		}
+
+		void flush()//flush all batches
+		{
+		  const auto queue_size = _queue.size();
+			if (queue_size == 0) return;
+
+			//handle batching
+			std::string batch, next;
+			_queue.pop(&batch); //there is at least one element
+
+			for (size_t i = 1; i < queue_size; ++i)
+			{
+				_queue.pop(&next);
+
+				size_t batch_size = batch.length() + next.length();
+				if (batch_size > _batch_max_size)
+				{
+          api_status status;
+					//the batch is about to reach the max size: send it
+					if(_sender.send(batch,&status) != error_code::success)
+					{
+            if (_perror_cb) _perror_cb->report_error(status);
+					}
+          
+					batch = next;
+				}
+				else
+					batch += "\n" + next;
+			}
+
+			//send remaining events
+			if (batch.size() > 0)
+			{
+        api_status status;
+        //the batch is about to reach the max size: send it
+        if (_sender.send(batch, &status) != error_code::success)
+        {
+          if (_perror_cb) _perror_cb->report_error(status);
+        }
+			}
+		}
+
+	public:
+		async_batcher(TSender& pipe, error_callback_fn* perror_cb = nullptr,
+                  size_t batch_max_size = (256 * 1024 - 1), size_t batch_timeout_ms = 1000, size_t queue_max_size = (8 * 1024))
+			: _sender(pipe),
+      _perror_cb(perror_cb),
+	    _batch_max_size(batch_max_size),
+			_batch_timeout_ms(batch_timeout_ms),
+			_queue_max_size(queue_max_size)
+		{
+			_thread_is_running = true;
+			_background_thread = std::thread(&async_batcher::timer, this);
```

Comment:
>_background_thread = std::thread(&async_batcher::timer, this); [](start = 3, length = 62)

can this throw?

Reply:
it can throw according to Yann. we need to handle

---
In reply to: [186743714](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186743714) [](ancestors = 186743714)